AuthTests/PublicKeyServiceTests.cs

```diff
@@ -0,0 +1,453 @@
+
+                ""issuer2"": ""{ \""keys\"": [{ \""kid\"": \""key2\"", \""kty\"": \""RSA\"", \""use\"": \""sig\"", \""n\"": \""another-modulus-value\"", \""e\"": \""another-exponent-value\"" }] }""
+
+            ";
+
+            mockHttpMessageHandler.Protected()
+
+              .Setup<Task<HttpResponseMessage>>("SendAsync", ItExpr.IsAny<HttpRequestMessage>(), ItExpr.IsAny<CancellationToken>())
+
+
+
+              .ReturnsAsync(new HttpResponseMessage()
+
+              {
+
+                  StatusCode = HttpStatusCode.OK,
+
+                  Content = new StringContent("invalid-json")
+
+              });
+
+
+
+            var httpClient = new HttpClient(mockHttpMessageHandler.Object);
+
+            var mockHttpClientFactory = new Mock<IHttpClientFactory>();
+
+            mockHttpClientFactory.Setup(factory
+
+              => factory.CreateClient("fdc-client")).Returns(httpClient);
+
+
+
+            var publicKeyService = new PublicKeyService(mockHttpClientFactory.Object, mockMemoryCache.Object);
+
+            var bffUrl = new Uri("https://dummy-bff/jwks");
+
+
+
+            // Act & Assert
+
+            await Assert.ThrowsAsync<JsonException>(() => publicKeyService.GetSigningKeysFromJwkAsync(bffUrl));
+
+        }
+
+
+
+
+
+        [Fact]
+
+        public async Task GetSigningKeysFromJwkAsync_HandlesException_AndThrows()
+
+        {
+
+            // Arrange
+
+            var mockMemoryCache = new Mock<IMemoryCache>();
+
+            var mockHttpMessageHandler = new Mock<HttpMessageHandler>();
+
+            mockHttpMessageHandler.Protected()
+
+              .Setup<Task<HttpResponseMessage>>("SendAsync", ItExpr.IsAny<HttpRequestMessage>(), ItExpr.IsAny<CancellationToken>())
+
+              .ThrowsAsync(new
+
+                HttpRequestException("Network error"));
+
+
+
+            var httpClient = new HttpClient(mockHttpMessageHandler.Object);
+
+            var mockHttpClientFactory = new Mock<IHttpClientFactory>();
+
+            mockHttpClientFactory.Setup(factory
+
+              => factory.CreateClient("fdc-client")).Returns(httpClient);
+
+
+
+            var publicKeyService = new PublicKeyService(mockHttpClientFactory.Object, mockMemoryCache.Object);
+
+            var bffUrl = new Uri("https://dummy-bff/jwks");
+
+
+
+            // Act & Assert
+
+            await Assert.ThrowsAsync<HttpRequestException>(() => publicKeyService.GetSigningKeysFromJwkAsync(bffUrl));
+
+        }
+
+
+
+
+
+        [Fact]
+
+        public async Task GetSigningKeysFromJwkAsync_ConcurrentRequests_OnlyOneFetchAndOthersUseCache()
+
+        {
+
+            // Arrange
+
+            var httpClient = HttpClient();
+
+            var mockHttpClientFactory = new Mock<IHttpClientFactory>();
+
+            mockHttpClientFactory.Setup(factory
+
+              => factory.CreateClient("fdc-client")).Returns(httpClient);
+
+            var memoryCache = new MemoryCache(new MemoryCacheOptions());
```

Comment:
can I pass this to simulate Semaphore

Reply:
I am not sure what is your real requirements. But seems like ICollectionFixture can be used to share between tests https://hamidmosalla.com/2020/02/02/xunit-part-5-share-test-context-with-iclassfixture-and-icollectionfixture/